django/db/backends/postgresql/schema.py

```diff
@@ -13,6 +13,16 @@ class DatabaseSchemaEditor(BaseDatabaseSchemaEditor):
     )
     sql_alter_sequence_type = "ALTER SEQUENCE IF EXISTS %(sequence)s AS %(type)s"
     sql_delete_sequence = "DROP SEQUENCE IF EXISTS %(sequence)s CASCADE"
+    sql_create_sequence = (
+        "CREATE SEQUENCE IF NOT EXISTS %(sequence)s "
+        "AS %(data_type)s "
+        "OWNED BY %(table)s.%(column)s"
+    )
+    sql_reset_sequence = (
+        "SELECT setval(pg_get_serial_sequence(%s, %s), "
+        "coalesce(max(%s), 1), max(%s) IS NOT NULL) FROM %s;"
+    )
+    sql_rename_sequence = "ALTER SEQUENCE %s RENAME TO %s;"
```

Comment:
Thinking more about it isn't there something fundamentally wrong here about not using the native `serial` types?

Why are we going the extra length of manually emulating through schema definition and explicitly using `nextval(pg_get_serial_sequence(...))` on `INSERT` what simply using `serial` and friends does?

I could be missing something here but it seems that most of the complexity of this patch stems from trying to emulate what Postgres does very well already?

Reply:
@charettes , it's because Postgres doesn't support it very well. `serial` is not a true type, and it's not possible to alter a field's type to `serial`. we have to emulate it because that's the only way it can work.

Reply:
Could you elaborate on how Postgres doesn't support `serial` very well? It has been used to power `AutoField` for over a decade before we switched to `IDENTITY` so I'm surprised to learn that it poses problems.

From my understanding the main reason for the manual sequence management and `db_default` assignment is to support altering back and forth between integer, auto, and serial fields. Is that really going to be a common enough use case that it warrants all the complexity it requires?

I think it's a question worth asking because there are no mentions of supporting this use case in the accepted ticket-27452 and I personally don't think that it will be common enough operation to warrant that much extra complexity to the schema editor.

Don't get me wrong, it is impressive that you got it working and I'm sure you learned a ton doing so. I'm just doubtful that the code complexity tradeoff is worth it (1.4k additional line) considering that this could be simply implemented using mixins including `db_returning = True` and using `get_db_type = "serial"` at first.

Reply:
You can use `serial` when creating a table or when adding a field, but you can't use it for anything else, e.g. you can't alter a column's type to serial and you can't cast to serial. But do let me know if I'm wrong, it didn't work when I tested it.

I agree it adds some complexity, but the majority of that 1.4k lines is just tests, the schema editor received 160 additions and 44 removals only. I assumed the functionality to transition between integer fields <-> serial fields <-> auto fields would be a requirement because of the similarities between these fields. If all we wanted to support was creating serial fields, I'm not sure this would need any changes to core, it could be a package.

Reply:
> From my understanding the main reason for the manual sequence management and db_default assignment is to support altering back and forth between integer, auto, and serial fields.

Yes this is correct.

Reply:
I think a migration story is important here - I'm sure there will be people who would have chosen `SerialField` if it was available and now want to migrate over.

@csirmazbendeguz can you share the raw SQL that didn't work when you tested? Maybe that will inspire a better solution.

Reply:
@LilyFoote , yes, I mean this one for example:

```sql
ALTER TABLE "schema_foo" ALTER COLUMN "field" TYPE bigserial USING "field"::bigserial
-- type "serial" does not exist
```